### PR TITLE
[Merged by Bors] - feat: the norm raised to a power is C^1

### PR DESCRIPTION
**Pull request by fpvandoorn**
* From the Sobolev inequality project
* Generalize `continuousAt_rpow_const`

Co-authored-by: Heather Macbeth 25316162+hrmacbeth@users.noreply.github.com

---
- [x] depends on: #14123 [one proof will fail until then]

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/InnerProductSpace/NormPow.lean [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace ℝ E]
+variable {F : Type*} [NormedAddCommGroup F] [NormedSpace ℝ F]
+
+theorem hasFDerivAt_norm_rpow (x : E) {p : ℝ} (hp : 1 < p) :
+    HasFDerivAt (fun x : E ↦ ‖x‖ ^ p) ((p * ‖x‖ ^ (p - 2)) • innerSL ℝ x) x := by
+  by_cases hx : x = 0
+  · simp [hx]
+    have h2p : 0 < p - 1 := sub_pos.mpr hp
+    rw [HasFDerivAt, hasFDerivAtFilter_iff_isLittleO]
+    simp [zero_lt_one.trans hp |>.ne']
```
Comment (sgouezel) [inline Mathlib/Analysis/InnerProductSpace/NormPow.lean:35]: nonterminal simp

File: Mathlib/Analysis/SpecialFunctions/Pow/Continuity.lean [hunk 2]
```diff
@@ -249,6 +249,9 @@ theorem continuousAt_rpow_const (x : ℝ) (q : ℝ) (h : x ≠ 0 ∨ 0 < q) :
   · exact (continuous_id'.prod_mk continuous_const).continuousAt
 #align real.continuous_at_rpow_const Real.continuousAt_rpow_const
 
+@[fun_prop]
+theorem continuous_rpow_const {q : ℝ} (h : 0 < q) : Continuous (fun x : ℝ => x ^ q) :=
```
Comment (sgouezel) [inline Mathlib/Analysis/SpecialFunctions/Pow/Continuity.lean:253]: Could you assume that `q` is nonnegative instead? (Should be trivial for `q = 0`).

File: Mathlib/Analysis/InnerProductSpace/NormPow.lean [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    ‖fderiv ℝ (fun x ↦ ‖f x‖ ^ (p : ℝ)) x‖₊ ≤ p * ‖f x‖₊ ^ ((p : ℝ) - 1) * ‖fderiv ℝ f x‖₊ :=
+  norm_fderiv_norm_rpow_le hf hp
+
+theorem contDiff_norm_rpow {p : ℝ} (hp : 1 < p) : ContDiff ℝ 1 (fun x : E ↦ ‖x‖ ^ p) := by
+  rw [contDiff_one_iff_fderiv]
+  refine ⟨fun x ↦ hasFDerivAt_norm_rpow x hp |>.differentiableAt, ?_⟩
+  simp_rw [continuous_iff_continuousAt]
+  intro x
+  by_cases hx : x = 0
+  · simp [hx, ContinuousAt, fderiv_norm_rpow (E := E) (x := 0) hp]
```
Comment (sgouezel) [inline Mathlib/Analysis/InnerProductSpace/NormPow.lean:107]: nonterminal simp.

**Review by sgouezel — COMMENTED** [review]
bors d+
Thanks!

File: Mathlib/Analysis/SpecialFunctions/Pow/Continuity.lean (hunk 2 above)
Comment (fpvandoorn) [reply Mathlib/Analysis/SpecialFunctions/Pow/Continuity.lean]: Thanks! I also generalized the previous lemma.

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**